### ISSUE TITLE
false positive non-exhaustive patterns: pattern guarded outside of match 

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#57247]
````rust
pub fn func(a: Option<i32>, b: Option<i32>) {
	if a.is_some() || b.is_some() {
		match &(a,b) {
/*
7 |         match &(a,b) {
  |               ^^^^^^ pattern `&(None, None)` not covered
 */
//			(None, None) => {} // will never match due to if-condition
			(Some(_a), None) => {}
			(None, Some(_b)) => {}
			(Some(_a), Some(_b)) => {}
		}
	}
}
````

We will only enter the match block if either `a` or `b` is some, thus `(None, None)` is redundant, however the compiler still requires it.

This also means that
````rust
pub fn func(a: Option<i32>, b: Option<i32>) {
	if a.is_some() || b.is_some() {
		match &(a,b) {
			(None, None) => { /* dead code */ } 
			(Some(_a), None) => {}
			(None, Some(_b)) => {}
			(Some(_a), Some(_b)) => {}
		}
	}
}
````
exhibits dead code which is not warned about currently.

meta:
````
rustc 1.33.0-nightly (9eac38634 2018-12-31)
binary: rustc
commit-hash: 9eac386342c601b14311b435f2b6d314fc817bb5
commit-date: 2018-12-31
host: x86_64-unknown-linux-gnu
release: 1.33.0-nightly
LLVM version: 8.0
````

**Comment by varkor** [comment]
We don't do any sort of value tracking at the moment: it's not specific to `match`. A very minimal example for `match` would be this (which doesn't produce any warnings):
```rust
fn main() {
    match true {
        true => {}
        false => {}
    }
}
```
